### DISCUSSION TITLE
验证安装样例运行后出现Fatal process OOM in insufficient memory to create an Isolate

### DISCUSSION THREAD

**Discussion by Amoukk** [microsoft/nni#3807] [Ideas]
## 问题概述

启动服务器的时候出现了Fatal process OOM in insufficient memory to create an Isolate 内存8G 启动前的时候内存占用也不超60% 但是最后就会弹出OOM
     
     OS: Windows64
     python version: 3.7.0

## 具体错误信息

```
E:\bc>nnictl create --config nni\examples\trials\mnist-tfv1\config_windows.yml
INFO: expand searchSpacePath: search_space.json to E:\bc\nni\examples\trials\mnist-tfv1\search_space.json
INFO: expand codeDir: . to E:\bc\nni\examples\trials\mnist-tfv1\.
INFO: Starting restful server...
ERROR: Restful server start failed!
INFO:  Stdout:
-----------------------------------------------------------------------
                Experiment start time 2020-04-20 08:15:40
-----------------------------------------------------------------------

<--- Last few GCs --->


<--- JS stacktrace --->





INFO:  Stderr:
-----------------------------------------------------------------------
                Experiment start time 2020-04-20 08:15:40
-----------------------------------------------------------------------

#
# Fatal process OOM in insufficient memory to create an Isolate
#


Traceback (most recent call last):
  File "d:\python\lib\site-packages\psutil\_pswindows.py", line 679, in wrapper
    return fun(self, *args, **kwargs)
  File "d:\python\lib\site-packages\psutil\_pswindows.py", line 933, in create_time
    user, system, created = cext.proc_times(self.pid)
ProcessLookupError: [Errno 3] No such process (originated from GetExitCodeProcess != STILL_ACTIVE)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "d:\python\lib\site-packages\psutil\__init__.py", line 373, in _init
    self.create_time()
  File "d:\python\lib\site-packages\psutil\__init__.py", line 723, in create_time
    self._create_time = self._proc.create_time()
  File "d:\python\lib\site-packages\psutil\_pswindows.py", line 681, in wrapper
    raise convert_oserror(err, pid=self.pid, name=self._name)
psutil.NoSuchProcess: psutil.NoSuchProcess process no longer exists (pid=16184)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "d:\python\lib\site-packages\nni_cmd\launcher.py", line 445, in launch_experiment
    kill_command(rest_process.pid)
  File "d:\python\lib\site-packages\nni_cmd\command_utils.py", line 37, in kill_command
    process = psutil.Process(pid=pid)
  File "d:\python\lib\site-packages\psutil\__init__.py", line 346, in __init__
    self._init(pid)
  File "d:\python\lib\site-packages\psutil\__init__.py", line 386, in _init
    raise NoSuchProcess(pid, None, msg)
psutil.NoSuchProcess: psutil.NoSuchProcess no process found with pid 16184

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "d:\python\lib\site-packages\nni_cmd\launcher.py", line 508, in create_experiment
    launch_experiment(args, experiment_config, 'new', config_file_name)
  File "d:\python\lib\site-packages\nni_cmd\launcher.py", line 447, in launch_experiment
    raise Exception(ERROR_INFO % 'Rest server stopped!')
Exception: ERROR: Rest server stopped!

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "d:\python\lib\site-packages\psutil\_pswindows.py", line 679, in wrapper
    return fun(self, *args, **kwargs)
  File "d:\python\lib\site-packages\psutil\_pswindows.py", line 933, in create_time
    user, system, created = cext.proc_times(self.pid)
ProcessLookupError: [Errno 3] No such process (originated from GetExitCodeProcess != STILL_ACTIVE)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "d:\python\lib\site-packages\psutil\__init__.py", line 373, in _init
    self.create_time()
  File "d:\python\lib\site-packages\psutil\__init__.py", line 723, in create_time
    self._create_time = self._proc.create_time()
  File "d:\python\lib\site-packages\psutil\_pswindows.py", line 681, in wrapper
    raise convert_oserror(err, pid=self.pid, name=self._name)
psutil.NoSuchProcess: psutil.NoSuchProcess process no longer exists (pid=16184)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "d:\python\lib\runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "d:\python\lib\runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "D:\python\Scripts\nnictl.exe\__main__.py", line 7, in <module>
  File "d:\python\lib\site-packages\nni_cmd\nnictl.py", line 223, in parse_args
    args.func(args)
  File "d:\python\lib\site-packages\nni_cmd\launcher.py", line 513, in create_experiment
    kill_command(restServerPid)
  File "d:\python\lib\site-packages\nni_cmd\command_utils.py", line 37, in kill_command
    process = psutil.Process(pid=pid)
  File "d:\python\lib\site-packages\psutil\__init__.py", line 346, in __init__
    self._init(pid)
  File "d:\python\lib\site-packages\psutil\__init__.py", line 386, in _init
    raise NoSuchProcess(pid, None, msg)
psutil.NoSuchProcess: psutil.NoSuchProcess no process found with pid 16184
```

**Comment by scarlett2018** [comment]
> ## 问题概述
> 启动服务器的时候出现了Fatal process OOM in insufficient memory to create an Isolate 内存8G 启动前的时候内存占用也不超60% 但是最后就会弹出OOM
> 
> ```
>  OS: Windows64
>  python version: 3.7.0
> ```
> 
> ## 具体错误信息
> ```
> E:\bc>nnictl create --config nni\examples\trials\mnist-tfv1\config_windows.yml
> INFO: expand searchSpacePath: search_space.json to E:\bc\nni\examples\trials\mnist-tfv1\search_space.json
> INFO: expand codeDir: . to E:\bc\nni\examples\trials\mnist-tfv1\.
> INFO: Starting restful server...
> ERROR: Restful server start failed!
> INFO:  Stdout:
> -----------------------------------------------------------------------
>                 Experiment start time 2020-04-20 08:15:40
> -----------------------------------------------------------------------
> 
> <--- Last few GCs --->
> 
> 
> <--- JS stacktrace --->
> 
> 
> 
> 
> 
> INFO:  Stderr:
> -----------------------------------------------------------------------
>                 Experiment start time 2020-04-20 08:15:40
> -----------------------------------------------------------------------
> 
> #
> # Fatal process OOM in insufficient memory to create an Isolate
> #
> 
> 
> Traceback (most recent call last):
>   File "d:\python\lib\site-packages\psutil\_pswindows.py", line 679, in wrapper
>     return fun(self, *args, **kwargs)
>   File "d:\python\lib\site-packages\psutil\_pswindows.py", line 933, in create_time
>     user, system, created = cext.proc_times(self.pid)
> ProcessLookupError: [Errno 3] No such process (originated from GetExitCodeProcess != STILL_ACTIVE)
> 
> During handling of the above exception, another exception occurred:
> 
> Traceback (most recent call last):
>   File "d:\python\lib\site-packages\psutil\__init__.py", line 373, in _init
>     self.create_time()
>   File "d:\python\lib\site-packages\psutil\__init__.py", line 723, in create_time
>     self._create_time = self._proc.create_time()
>   File "d:\python\lib\site-packages\psutil\_pswindows.py", line 681, in wrapper
>     raise convert_oserror(err, pid=self.pid, name=self._name)
> psutil.NoSuchProcess: psutil.NoSuchProcess process no longer exists (pid=16184)
> 
> During handling of the above exception, another exception occurred:
> 
> Traceback (most recent call last):
>   File "d:\python\lib\site-packages\nni_cmd\launcher.py", line 445, in launch_experiment
>     kill_command(rest_process.pid)
>   File "d:\python\lib\site-packages\nni_cmd\command_utils.py", line 37, in kill_command
>     process = psutil.Process(pid=pid)
>   File "d:\python\lib\site-packages\psutil\__init__.py", line 346, in __init__
>     self._init(pid)
>   File "d:\python\lib\site-packages\psutil\__init__.py", line 386, in _init
>     raise NoSuchProcess(pid, None, msg)
> psutil.NoSuchProcess: psutil.NoSuchProcess no process found with pid 16184
> 
> During handling of the above exception, another exception occurred:
> 
> Traceback (most recent call last):
>   File "d:\python\lib\site-packages\nni_cmd\launcher.py", line 508, in create_experiment
>     launch_experiment(args, experiment_config, 'new', config_file_name)
>   File "d:\python\lib\site-packages\nni_cmd\launcher.py", line 447, in launch_experiment
>     raise Exception(ERROR_INFO % 'Rest server stopped!')
> Exception: ERROR: Rest server stopped!
> 
> During handling of the above exception, another exception occurred:
> 
> Traceback (most recent call last):
>   File "d:\python\lib\site-packages\psutil\_pswindows.py", line 679, in wrapper
>     return fun(self, *args, **kwargs)
>   File "d:\python\lib\site-packages\psutil\_pswindows.py", line 933, in create_time
>     user, system, created = cext.proc_times(self.pid)
> ProcessLookupError: [Errno 3] No such process (originated from GetExitCodeProcess != STILL_ACTIVE)
> 
> During handling of the above exception, another exception occurred:
> 
> Traceback (most recent call last):
>   File "d:\python\lib\site-packages\psutil\__init__.py", line 373, in _init
>     self.create_time()
>   File "d:\python\lib\site-packages\psutil\__init__.py", line 723, in create_time
>     self._create_time = self._proc.create_time()
>   File "d:\python\lib\site-packages\psutil\_pswindows.py", line 681, in wrapper
>     raise convert_oserror(err, pid=self.pid, name=self._name)
> psutil.NoSuchProcess: psutil.NoSuchProcess process no longer exists (pid=16184)
> 
> During handling of the above exception, another exception occurred:
> 
> Traceback (most recent call last):
>   File "d:\python\lib\runpy.py", line 193, in _run_module_as_main
>     "__main__", mod_spec)
>   File "d:\python\lib\runpy.py", line 85, in _run_code
>     exec(code, run_globals)
>   File "D:\python\Scripts\nnictl.exe\__main__.py", line 7, in <module>
>   File "d:\python\lib\site-packages\nni_cmd\nnictl.py", line 223, in parse_args
>     args.func(args)
>   File "d:\python\lib\site-packages\nni_cmd\launcher.py", line 513, in create_experiment
>     kill_command(restServerPid)
>   File "d:\python\lib\site-packages\nni_cmd\command_utils.py", line 37, in kill_command
>     process = psutil.Process(pid=pid)
>   File "d:\python\lib\site-packages\psutil\__init__.py", line 346, in __init__
>     self._init(pid)
>   File "d:\python\lib\site-packages\psutil\__init__.py", line 386, in _init
>     raise NoSuchProcess(pid, None, msg)
> psutil.NoSuchProcess: psutil.NoSuchProcess no process found with pid 16184
> ```

Hi @Amou6845 谢谢关注NNI并试用examples，请问你的其他NNI环境信息是什么：

**nni Environment:**
nni version:
nni mode(local|pai|remote):
OS: Windows64
python version:  3.7.0
is conda or virtualenv used?:
is running in docker?:

**Are you able to reproduce? if so, how to reproduce?**

**Comment by Amoukk** [comment]
> Hi @Amou6845 谢谢关注NNI并试用examples，请问你的其他NNI环境信息是什么：
> 
> **nni Environment:**
> nni version:
> nni mode(local|pai|remote):
> OS: Windows64
> python version: 3.7.0
> is conda or virtualenv used?:
> is running in docker?:
> 
> **Are you able to reproduce? if so, how to reproduce?**

**nni Environment:**
nni version:v1.5
nni mode(local|pai|remote):local
OS: Windows64
python version: 3.7.0
is conda or virtualenv used?:no
is running in docker?:no

目前还没有找到能够解决的办法 还请指导一下如何解决

**Comment by liuzhe-lz** [comment]
请先跟踪一下内存占用情况，确认占用大量内存的是哪个进程。

**Comment by Amoukk** [comment]
> 
> 
> 请先跟踪一下内存占用情况，确认占用大量内存的是哪个进程。

已经尝试过关闭占用较大的进程 让内存保持在占用率50%左右 但是启动服务器仍然出现OOM情况 启动过程也没有内存占用率的波动

**Comment by liuzhe-lz** [comment]
也就是说启动过程中内存并没有占满但是node.js报OOM了？

**Comment by Amoukk** [comment]
> 
> 
> 也就是说启动过程中内存并没有占满但是node.js报OOM了？

是的 在任务管理器中并没有看见内存占用率变化

**Comment by liuzhe-lz** [comment]
这个看起来比较像运行环境的问题，我们在自己的机器上没有能够复现。
如果方便的话请常试一下在[WSL](https://docs.microsoft.com/en-us/windows/wsl/install-win10)里面运行Linux版NNI。

**Comment by SkyloveQiu** [comment]
我也有同样的问题，请问有人解决了吗。

**Comment by scarlett2018** [comment]
@Amoukk @SkyloveQiu - are you still seeing the problem? thanks.

**Comment by SkyloveQiu** [comment]
@scarlett2018 We decided to use WSL for that, the problem was resolved by using wsl
